File: app/src/main/java/com/udacity/notepad/data/NoteDatabase.java
package com.udacity.notepad.data;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import static android.provider.BaseColumns._ID;

public class NoteDatabase {

    private final NotesOpenHelper helper;

    public NoteDatabase(Context context) {
        helper = new NotesOpenHelper(context);
    }

    private static Note fromCursor(Cursor cursor) {
        int col = 0;
        Note note = new Note();
        note.setId(cursor.getInt(col++));
        note.setText(cursor.getString(col++));
        note.setPinned(cursor.getInt(col++) != 0);
        note.setCreatedAt(new Date(cursor.getLong(col++)));
        note.setUpdatedAt(new Date(cursor.getLong(col)));
        return note;
    }

    private static List<Note> allFromCursor(Cursor cursor) {
        List<Note> retval = new ArrayList<>();
        while (cursor.moveToNext()) {
            retval.add(fromCursor(cursor));
        }
        return retval;
    }

    private static ContentValues fromNote(Note note) {
        ContentValues values = new ContentValues();
        int id = note.getId();
        if (id != -1) {
            values.put(_ID, id);
        }
        values.put(Companion.getTEXT(), note.getText());
        values.put(Companion.getIS_PINNED(), note.isPinned());
        values.put(Companion.getCREATED_AT(), note.getCreatedAt().getTime());
        values.put(Companion.getUPDATED_AT(), note.getUpdatedAt().getTime());
        return values;
    }

    private static List<ContentValues> fromNotes(Note[] notes) {
        List<ContentValues> values = new ArrayList<>();
        for (Note note : notes) {
            values.add(fromNote(note));
        }
        return values;
    }

    public List<Note> getAll() {
        Cursor cursor = helper.getReadableDatabase().query(Companion.get_TABLE_NAME(),
                null,
                null,
                null,
                null,
                null,
                Companion.getCREATED_AT());
        List<Note> retval = allFromCursor(cursor);
        cursor.close();
        return retval;
    }

    public List<Note> loadAllByIds(int... ids) {
        StringBuilder questionMarks = new StringBuilder();
        int i = 0;
        while (i++ < ids.length) {
            questionMarks.append("?");
            if (i <= ids.length - 1) {
                questionMarks.append(", ");
            }
        }
        String[] args = new String[ids.length];
        for (i = 0; i < ids.length; ++i) {
            args[i] = Integer.toString(ids[i]);
        }
        String selection = _ID + " IN (" + questionMarks + ")";
        Cursor cursor = helper.getReadableDatabase().query(Companion.get_TABLE_NAME(),
                null,
                selection,
                args,
                null,
                null,
                Companion.getCREATED_AT());
        List<Note> retval = allFromCursor(cursor);
        cursor.close();
        return retval;
    }

    public void insert(Note... notes) {
        List<ContentValues> values = fromNotes(notes);
        SQLiteDatabase db = helper.getWritableDatabase();
        db.beginTransaction();
        try {
            for (ContentValues value : values) {
                db.insert(Companion.get_TABLE_NAME(), null, value);
            }
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }
    }

    public void update(Note note) {
        ContentValues values = fromNote(note);
        helper.getWritableDatabase().update(Companion.get_TABLE_NAME(),
                values,
                _ID + " = ?",
                new String[]{ Integer.toString(note.getId()) });
    }

    public void delete(Note note) {
        helper.getWritableDatabase().delete(Companion.get_TABLE_NAME(),
                _ID + " = ?",
                new String[]{ Integer.toString(note.getId()) });
    }
}

File: app/src/main/java/com/udacity/notepad/data/NoteDatabase.kt
package com.udacity.notepad.data

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.provider.BaseColumns._ID
import com.udacity.notepad.data.NotesContract.NoteTable.CREATED_AT
import com.udacity.notepad.data.NotesContract.NoteTable.IS_PINNED
import com.udacity.notepad.data.NotesContract.NoteTable.TEXT
import com.udacity.notepad.data.NotesContract.NoteTable.UPDATED_AT
import com.udacity.notepad.data.NotesContract.NoteTable._TABLE_NAME
import org.jetbrains.anko.db.transaction
import java.util.*

class NoteDatabase(context: Context) {

    private val helper: NotesOpenHelper = NotesOpenHelper(context)

    fun getAll(): List<Note> {
            val cursor = helper.readableDatabase.query(
                    _TABLE_NAME,
                    null,
                    null,
                    null,
                    null,
                    null,
                    CREATED_AT)

        return cursor.use(this::allFromCursor)
        }

    private fun fromCursor(cursor: Cursor): Note {
        var col = 0
        return Note().apply {
            id = cursor.getInt(col++)
            text = cursor.getString(col++)
            isPinned = cursor.getInt(col++) != 0
            createdAt = Date(cursor.getLong(col++))
            updatedAt = Date(cursor.getLong(col))
        }
    }

    private fun allFromCursor(cursor: Cursor): List<Note> {
        val retval = ArrayList<Note>()
        while (cursor.moveToNext()) {
            retval.add(fromCursor(cursor))
        }
        return retval
    }

    private fun fromNote(note: Note): ContentValues {
        return ContentValues().apply {
            val noteId = note.id
            if (noteId != -1) {
                put(_ID, noteId)
            }
            put(TEXT, note.text)
            put(IS_PINNED, note.isPinned)
            put(CREATED_AT, note.createdAt.time)
            put(UPDATED_AT, note.updatedAt!!.time)
        }
    }

    private fun fromNotes(notes: Array<out Note>): List<ContentValues> {
        val values = ArrayList<ContentValues>()
        for (note in notes) {
            values.add(fromNote(note))
        }
        return values
    }

    fun loadAllByIds(vararg ids: Int): List<Note> {
        val questionMarks = ids.map { "?" }.joinToString { ", " }
        val args = ids.map { it.toString() }
        val selection = "$_ID + IN ($questionMarks)"
        val cursor = helper.readableDatabase.query(_TABLE_NAME,
                null,
                selection,
                args.toTypedArray(),
                null,
                null,
                CREATED_AT)
        return cursor.use(this::allFromCursor)
    }

    fun insert(vararg notes: Note) {
        val values = fromNotes(notes)
        val db = helper.writableDatabase
        db.transaction {
            for (value in values) {
                db.insert(_TABLE_NAME, null, value)
            }
        }
    }

    fun update(note: Note) {
        val values = fromNote(note)
        helper.writableDatabase.update(_TABLE_NAME,
                values,
                _ID + " = ?",
                arrayOf(Integer.toString(note.id)))
    }

    fun delete(note: Note) {
        helper.writableDatabase.delete(_TABLE_NAME,
                _ID + " = ?",
                arrayOf(Integer.toString(note.id)))
    }
}
